urla958522fc1_spirosikmd_inmdsy_08_tgz-pJ8-TesterAppJ8/src/nl/rug/peerbox/logic/filesystem/FileServer.java

```diff
@@ -0,0 +1,102 @@
+package nl.rug.peerbox.logic.filesystem;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import nl.rug.peerbox.logic.Context;
+import nl.rug.peerbox.logic.Peerbox;
+
+import org.apache.log4j.Logger;
+
+public final class FileServer implements Runnable {
+
+	
+
+	private static final Logger logger = Logger.getLogger(FileServer.class);
+	private final ExecutorService pool = Executors.newFixedThreadPool(5);
+	private final Context ctx;
+	boolean alive = true;
+	
+	public FileServer(Context ctx) {
+		this.ctx = ctx;
+	}
+	
+	@Override
+	public void run() {
+		logger.info("Starting server " + ctx.getLocalPeer().getPort());
+		try (ServerSocket server = new ServerSocket()) {
+			server.bind(new InetSocketAddress(ctx.getLocalPeer().getAddress(), ctx.getLocalPeer().getPort()));
+			while (!Thread.currentThread().isInterrupted()) {
+				logger.info("Waiting for incoming connection");
+				try {
+					final Socket s = server.accept();
+					logger.debug("Accepted incoming connection from "
+							+ s.getRemoteSocketAddress());
+					pool.execute(new SendFileTask(s));
+				} catch (IOException e) {
+					logger.error(e);
+				}
+			}
+			pool.shutdownNow();
+
+		} catch (IOException e) {
+			logger.error(e);
+		}
+
+	}
+
+	private final class SendFileTask implements Runnable {
+		private final Socket s;
+		private final Context ctx;
+
+		private SendFileTask(Socket s) {
+			this.s = s;
+			this.ctx = Peerbox.getInstance();
+		}
+
+		@Override
+		public void run() {
+			try {
+				BufferedReader st = new BufferedReader(new InputStreamReader(
+						s.getInputStream()));
+				String fileid = st.readLine();
+				logger.info("File " + fileid + " has been requested.");
+				File myFile = new File(ctx.getPathToPeerbox()
+						+ System.getProperty("file.separator") + fileid);
+				byte[] mybytearray = new byte[(int) myFile.length()];
+
+				BufferedInputStream bis = new BufferedInputStream(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *bis*. The resource is referenced at lines 78, 81. The resource is closed at line 86. There are other execution paths that do not contain closure statements, for example, when OutputStream.write throws an exception. Ensure *bis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).